x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java

```diff
@@ -493,6 +541,16 @@ private Map<DiscoveryNode, NodeLoad> detectNodeLoads(List<DiscoveryNode> nodes,
             );
     }
 
+    private boolean isScalingPossible(List<DiscoveryNode> nodes) {
+        OptionalLong smallestMLNode = nodes.stream().map(NodeLoadDetector::getNodeSize).flatMapToLong(OptionalLong::stream).min();
+
+        // We can scale horizontally
+        return maxLazyMLNodes > nodes.size()
+            // We can scale vertically
+            // TODO this currently only considers memory. We should also consider CPU when autoscaling by CPU is possible.
```

Comment:
I think this TODO comment should be changed, as it is a bit misleading at the moment.

If we can scale up memory then we'll also scale up CPU at the same time, so the current check is fine for whether scaling is possible at all.

What is missing (and has been missing since day 1 of memory-only autoscaling) is taking account of the possibility that although we can scale up we cannot scale up to the level required.  This is more likely to happen with CPU than with memory.  For example, we might be 32GB off of the maximum ML tier size and someone wants to start a 45GB job.  As this code stands we'll scale up to maximum size then find we still cannot start that job.  In the CPU case we might be 8 CPUs off of the maximum ML tier size and somebody wants to allocate a model 100 times with 4 threads per allocation.  However, in the CPU case it's not as bad to scale up to the maximum possible even though it doesn't satisfy what was requested, because we'll still allocate what we can.

Please adjust the comment so that somebody doesn't make the change implied by the current TODO without thinking about the effects.